src/main/java/fi/helsinki/ubipositioning/trilateration/LocationService.java

```diff
@@ -67,34 +105,41 @@
 
         NonLinearLeastSquaresSolver solver;
         solver = new NonLinearLeastSquaresSolver(new TrilaterationFunction(positions, distances), new LevenbergMarquardtOptimizer());
-        return solver.solve();
-    }
+        LeastSquaresOptimizer.Optimum optimum = solver.solve();
+
+        // Location in form of [x,y,z].
+        double[] centroid = optimum.getPoint().toArray();
+
+        // Standard error in form of [x,y,z].
+        double[] standardDeviation = optimum.getSigma(0).toArray();
+
+        // Covariance matrix.
+        RealMatrix covMatrix = optimum.getCovariances(0);
 
-    public void setCalculateDistance(boolean calculateDistance) {
-        this.calculateDistance = calculateDistance;
+        // Convert result into proper model.
+        return resultConverter.convert(beacon, centroid, standardDeviation, covMatrix);
     }
 
     /**
-     * Converts rssi value into distance of millimeters.
+     * Calculates location of all the devices. Ignoring those that can't be solved.
      *
-     * @param rssi signal strength that BLE listener has picked up..
-     * @param measuredPower Base power of BLE listener.
+     * @param beacons List of devices that needs to be located.
      *
-     * @return distance in millimeters of device which rssi is given.
+     * @see LocationService#calculateLocation(Beacon)
+     *
+     * @return Location of all the devices that can be pinpointed using trilateration.
      */
-    public double getDistanceFromRssi(double rssi, double measuredPower) {
-        if (this.calculateDistance) {
-            return Math.pow(10, (measuredPower - rssi) / (10 * this.environmentalFactor)) * 1000;
-        } else {
-            return rssi - measuredPower;
-        }
-    }
+    @Override
+    public List<Location> calculateAllLocations(List<Beacon> beacons) {
+        List<Location> list = new ArrayList<>();
 
-    public int getEnvironmentalFactor() {
-        return environmentalFactor;
-    }
+        beacons.forEach(beacon -> {
+            try {
+                list.add(calculateLocation(beacon));
+            } catch (Exception ignored) {
```

Comment:
Any chance we could catch more specific exception type? If i understand this right, this will catch also this exception `throw new IllegalArgumentException("BLE device must have been seen by at least one observer!");` , which might not be what we want.